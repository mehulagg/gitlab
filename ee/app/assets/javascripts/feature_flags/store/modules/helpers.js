import _ from 'underscore';
import {
  ROLLOUT_STRATEGY_ALL_USERS,
  ROLLOUT_STRATEGY_PERCENT_ROLLOUT,
  INTERNAL_ID_PREFIX,
  DEFAULT_PERCENT_ROLLOUT,
  PERCENT_ROLLOUT_GROUP_ID,
} from '../../constants';

/**
 * Converts raw scope objects fetched from the API into an array of scope
 * objects that is easier/nicer to bind to in Vue.
 * @param {Array} scopesFromRails An array of scope objects fetched from the API
 */
export const mapToScopesViewModel = scopesFromRails =>
  (scopesFromRails || []).map(s => {
    const [strategy] = s.strategies || [];

    const rolloutStrategy = strategy ? strategy.name : ROLLOUT_STRATEGY_ALL_USERS;

    let rolloutPercentage = DEFAULT_PERCENT_ROLLOUT;
    if (strategy && strategy.parameters && strategy.parameters.percentage) {
      rolloutPercentage = strategy.parameters.percentage;
    }

    return {
      id: s.id,
      environmentScope: s.environment_scope,
      active: Boolean(s.active),
      canUpdate: Boolean(s.can_update),
      protected: Boolean(s.protected),
      rolloutStrategy,
      rolloutPercentage,

      // eslint-disable-next-line no-underscore-dangle
      shouldBeDestroyed: Boolean(s._destroy),
    };
  });

/**
 * Converts the parameters emitted by the Vue component into
 * the shape that the Rails API expects.
 * @param {Array} scopesFromVue An array of scope objects from the Vue component
 */
export const mapFromScopesViewModel = params => {
  const scopes = (params.scopes || []).map(s => {
    const parameters = {};
    if (s.rolloutStrategy === ROLLOUT_STRATEGY_PERCENT_ROLLOUT) {
      parameters.groupId = PERCENT_ROLLOUT_GROUP_ID;
      parameters.percentage = s.rolloutPercentage;
    }

    // Strip out any internal IDs
    const id = _.isString(s.id) && s.id.startsWith(INTERNAL_ID_PREFIX) ? undefined : s.id;

    return {
      id,
      environment_scope: s.environmentScope,
      active: s.active,
      can_update: s.canUpdate,
      protected: s.protected,
      _destroy: s.shouldBeDestroyed,
      strategies: [
        {
          name: s.rolloutStrategy,
          parameters,
        },
      ],
    };
  });

  return {
    operations_feature_flag: {
      name: params.name,
      description: params.description,
      scopes_attributes: scopes,
    },
  };
};

/**
 * Creates a new feature flag environment scope object for use
 * in a Vue component.  An optional parameter can be passed to
 * override the property values that are created by default.
 *
 * @param {Object} overrides An optional object whose
 * property values will be used to override the default values.
 *
 */
export const createNewEnvironmentScope = (overrides = {}) => {
  const defaultScope = {
    environmentScope: '',
    active: false,
    id: _.uniqueId(INTERNAL_ID_PREFIX),
    rolloutStrategy: ROLLOUT_STRATEGY_ALL_USERS,
    rolloutPercentage: DEFAULT_PERCENT_ROLLOUT,
  };

  const newScope = {
    ...defaultScope,
    ...overrides,
  };

  if (gon && gon.features && gon.features.featureFlagPermissions) {
    newScope.canUpdate = true;
    newScope.protected = false;
  }

  return newScope;
};
