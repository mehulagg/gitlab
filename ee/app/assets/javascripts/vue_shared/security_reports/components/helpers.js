import { EMPTY_BODY_MESSAGE } from './constants';

export const bodyWithFallBack = (body) => {
  // Return string if defined or blank string, otherwise fallback
  return body === '' ? '' : body || EMPTY_BODY_MESSAGE;
};
