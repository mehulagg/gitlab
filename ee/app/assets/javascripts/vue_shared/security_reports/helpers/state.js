export default {
  sast: {
    isLoading: false,
    hasError: false,
    newIssues: [],
    resolvedIssues: [],
    allIssues: [],
  },
  sastContainer: {
    approved: [],
    unapproved: [],
    vulnerabilities: [],
  },
  dast: [],
  codeclimate: {
    newIssues: [],
    resolvedIssues: [],
  },
};
