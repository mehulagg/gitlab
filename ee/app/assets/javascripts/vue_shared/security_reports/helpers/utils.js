import { stripHtml } from '~/lib/utils/text_utility';

/**
 * Parses SAST and Codeclimate Issues into a common and reusable format
 * to reuse the same vue component.
 * [
 *   {
 *     name: String,
 *     priority: String,
 *     fingerprint: String,
 *     path: String,
 *     line: Number,
 *     urlPath: String
 *   }
 * ]
 * @param {array} issues
 * @return {array}
 */
export const parseIssues = (issues = [], path = '') => issues.map((issue) => {
  const parsedIssue = {
    name: issue.description || issue.message,
    ...issue,
  };

  // code quality
  if (issue.location) {
    let parseCodeQualityUrl;

    if (issue.location.path) {
      parseCodeQualityUrl = `${path}/${issue.location.path}`;
      parsedIssue.path = issue.location.path;
    }

    if (issue.location.lines && issue.location.lines.begin) {
      parsedIssue.line = issue.location.lines.begin;
      parseCodeQualityUrl += `#L${issue.location.lines.begin}`;
    }

    parsedIssue.urlPath = parseCodeQualityUrl;

  // security
  } else if (issue.file) {
    let parsedSecurityUrl = `${path}/${issue.file}`;
    parsedIssue.path = issue.file;

    if (issue.line) {
      parsedSecurityUrl += `#L${issue.line}`;
    }
    parsedIssue.urlPath = parsedSecurityUrl;
  }

  return parsedIssue;
});

/**
 * Compares two arrays by the given key and returns the difference
 *
 * @param {Array} firstArray
 * @param {Array} secondArray
 * @param {String} key
 * @returns {Array}
 */
export const filterByKey = (firstArray = [], secondArray = [], key = '') => firstArray
  .filter(item => !secondArray.find(el => el[key] === item[key]));

/**
 * Parses DAST results into a common format to allow to use the same Vue component
 * And adds an external link
 *
 * @param {Array} data
 * @returns {Array}
 */
export const parseSastContainer = (data = []) => data.map(el => ({
  name: el.vulnerability,
  priority: el.severity,
  path: el.namespace,
  // external link to provide better description
  nameLink: `https://cve.mitre.org/cgi-bin/cvename.cgi?name=${el.vulnerability}`,
  ...el,
}));

/**
 * Utils functions to set the reports
 */

/**
 * Compares sast results and returns the formatted report
 *
 * Security report has 3 types of issues, newIssues, resolvedIssues and allIssues.
 *
 * When we have both base and head:
 * - newIssues = head - base
 * - resolvedIssues = base - head
 * - allIssues = head - newIssues - resolvedIssues
 *
 * When we only have head
 * - newIssues = head
 * - resolvedIssues = 0
 * - allIssues = 0
 * @param {*} data
 * @returns {Object}
 */
export const setSastReport = (data = {}) => {
  const securityReport = {};

  if (data.base) {
    const filterKey = 'cve';
    const parsedHead = parseIssues(data.head, data.headBlobPath);
    const parsedBase = parseIssues(data.base, data.baseBlobPath);

    securityReport.newIssues = filterByKey(
      parsedHead,
      parsedBase,
      filterKey,
    );
    securityReport.resolvedIssues = filterByKey(
      parsedBase,
      parsedHead,
      filterKey,
    );

    // Remove the new Issues and the added issues
    securityReport.allIssues = filterByKey(
      parsedHead,
      securityReport.newIssues.concat(securityReport.resolvedIssues),
      filterKey,
    );
  } else {
    securityReport.newIssues = parseIssues(data.head, data.headBlobPath);
  }

  return securityReport;
};

export const setSastContainerReport = (data = {}) => {
  const unapproved = data.unapproved || [];

  const parsedVulnerabilities = parseSastContainer(data.vulnerabilities);

  // Approved can be calculated by subtracting unapproved from vulnerabilities.
  return {
    vulnerabilities: parsedVulnerabilities || [],
    approved: parsedVulnerabilities
    .filter(item => !unapproved.find(el => el === item.vulnerability)) || [],
    unapproved: parsedVulnerabilities
    .filter(item => unapproved.find(el => el === item.vulnerability)) || [],
  };
};

/**
 * Dast Report sends some keys in HTML, we need to strip the `<p>` tags.
 * This should be moved to the backend.
 *
 * @param {Array} data
 * @returns {Array}
 */
export const setDastReport = data => data.site.alerts.map(alert => ({
  name: alert.name,
  parsedDescription: stripHtml(alert.desc, ' '),
  priority: alert.riskdesc,
  ...alert,
}));
