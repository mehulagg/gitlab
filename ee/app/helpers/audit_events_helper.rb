# frozen_string_literal: true

module AuditEventsHelper
  def human_text(details)
    ::Audit::Details.humanize(details)
  end
end
