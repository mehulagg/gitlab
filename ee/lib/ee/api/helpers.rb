module EE
  module API
    module Helpers
      extend ::Gitlab::Utils::Override

      override :current_user
      def current_user
        strong_memoize(:current_user) do
          user = super

          if user
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :user, user.id)
          end

          user
        end
      end

      def check_project_feature_available!(feature)
        not_found! unless user_project.feature_available?(feature)
      end
    end
  end
end
