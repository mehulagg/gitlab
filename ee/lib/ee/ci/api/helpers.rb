module EE
  module Ci
    module API
      module Helpers
        def authenticate_build!
          id = params[:id]

          if id
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :build, id)
          end

          super
        end

        def current_runner
          token = params[:token]

          if token
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :runner, token)
          end

          super
        end
      end
    end
  end
end
