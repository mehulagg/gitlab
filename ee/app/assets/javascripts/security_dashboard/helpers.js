import isPlainObject from 'lodash/isPlainObject';
import { ALL, BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { s__, __ } from '~/locale';

const parseOptions = obj =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = projects =>
  projects.map(p => ({ id: p.id.split('/').pop(), name: p.name }));

export const stateFilter = {
  name: s__('SecurityReports|Status'),
  id: 'state',
  allOption: { id: ALL, name: s__('VulnerabilityStatusTypes|All') },
  options: parseOptions(VULNERABILITY_STATES),
};

export const severityFilter = {
  name: s__('SecurityReports|Severity'),
  id: 'severity',
  allOption: BASE_FILTERS.severity,
  options: parseOptions(SEVERITY_LEVELS),
};

const getScannerFilterOptions = () => {
  const options = parseOptions(REPORT_TYPES);
  return options.map(option => ({
    ...option,
    queryId: `GitLab:${option.id}`,
    externalId: 'NO_SCANNERS_FOUND',
  }));
};

export const scannerFilter = {
  name: s__('Reports|Scanner'),
  id: 'reportType',
  allOption: BASE_FILTERS.report_type,
  options: getScannerFilterOptions(),
};

export const getProjectFilter = projects => {
  if (!projects?.length) return undefined;

  return {
    name: s__('SecurityReports|Project'),
    id: 'projectId',
    allOption: BASE_FILTERS.project_id,
    options: mapProjects(projects),
  };
};

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter(entry => entry !== null);
};

/**
 * We have disabled loading hasNextPage from GraphQL as it causes timeouts in database,
 * instead we have to calculate that value based on the existence of endCursor. When endCursor
 * is empty or has null value, that means that there is no next page to be loaded from GraphQL API.
 *
 * @param {Object} pageInfo
 * @returns {Object}
 */
export const preparePageInfo = pageInfo => {
  return { ...pageInfo, hasNextPage: Boolean(pageInfo?.endCursor) };
};

export const createProjectLoadingError = () => __('An error occurred while retrieving projects.');

export default () => ({});
