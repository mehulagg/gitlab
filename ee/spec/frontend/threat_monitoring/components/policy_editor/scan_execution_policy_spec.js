import BasePolicy from 'ee/threat_monitoring/components/policy_editor/base_policy.vue';
import toYaml from 'ee/threat_monitoring/components/policy_editor/lib/to_yaml';
import ScanExecutionPolicy from 'ee/threat_monitoring/components/policy_editor/scan_execution_policy.vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';

describe('ScanExecutionPolicy component', () => {
  let wrapper;
  const policy = {
    name: 'test-policy',
    description: 'test description',
    endpointLabels: '',
    rules: [],
  };
  const unsupportedYaml = 'unsupportedPrimaryKey: test';

  const findDescription = () => wrapper.findByTestId('description');

  const factory = ({ propsData } = {}) => {
    wrapper = shallowMountExtended(ScanExecutionPolicy, {
      propsData: {
        ...propsData,
      },
      stubs: {
        BasePolicy,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('supported YAML', () => {
    beforeEach(() => {
      factory({ propsData: { value: toYaml(policy) } });
    });

    it('renders policy preview tabs', () => {
      expect(wrapper.find('div').element).toMatchSnapshot();
    });

    it('does render the policy description', () => {
      expect(findDescription().exists()).toBe(true);
      expect(findDescription().text()).toContain('test description');
    });
  });

  describe('unsupported YAML', () => {
    beforeEach(() => {
      factory({ propsData: { value: unsupportedYaml } });
    });

    it('renders policy preview tabs', () => {
      expect(wrapper.find('div').element).toMatchSnapshot();
    });

    it('does not render the policy description', () => {
      expect(findDescription().exists()).toBe(false);
    });
  });
});
