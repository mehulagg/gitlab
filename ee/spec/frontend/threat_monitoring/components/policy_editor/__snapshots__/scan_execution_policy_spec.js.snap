// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanExecutionPolicy component supported YAML renders policy preview tabs 1`] = `
<div
  value="apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
description: test description
metadata:
  name: test-policy
spec:
  endpointSelector:
    matchLabels:
      network-policy.gitlab.com/disabled_by: gitlab
"
>
  <h5
    class="gl-mt-3"
  >
    Type
  </h5>
   
  <p>
    Scan execution
  </p>
   
  <div>
    <policy-info-row-stub
      data-testid="description"
      label="Description"
    >
      test description
    </policy-info-row-stub>
     
    <!---->
     
    <!---->
     
    <policy-info-row-stub
      label="Enforcement status"
    >
      Disabled
    </policy-info-row-stub>
     
    <!---->
  </div>
</div>
`;

exports[`ScanExecutionPolicy component unsupported YAML renders policy preview tabs 1`] = `
<div
  value="unsupportedPrimaryKey: test"
>
  <h5
    class="gl-mt-3"
  >
    Type
  </h5>
   
  <p>
    Scan execution
  </p>
   
</div>
`;
